Kernel/LPOComparator.hpp

```diff
@@ -39,43 +39,24 @@ class LPOComparator
     T_UNKNOWN,
   };
 
-  struct Node;
+  class Node;
 
   struct Branch {
     BranchTag tag;
-    Node* n;
+    std::shared_ptr<Node> n;
```

Comment:
I only said it looked like a `shared_ptr`! Did you check for overhead? I think `shared_ptr<T>` has both a strong and weak count and maybe has to use `size_t` for counts rather than `unsigned`. If the overhead is unacceptable we could also look at `Lib::SmartPtr`.

But the code is much nicer. :innocent:

Reply:
I agree that it's much nicer! I assume there is just a little bit of overhead during `extend` calls. Maybe @quickbeam123 can do one more test with lpo? 🙂

Reply:
Perhaps this is the usual [`std::make_shared` footgun](https://stackoverflow.com/questions/20895648/difference-in-make-shared-and-normal-shared-ptr-in-c?rq=3) biting us?

Reply:
Maybe I was a bit too fast to change to `std::shared_ptr`. I can try the `Lib::SharedPtr` variant.

Reply:
Sorry for doing the lazy thing. I've just changed it to `Lib::SmartPtr`, but only did some local tests which show no difference in performance. @quickbeam123 can we try again with this version?

Reply:
Suspicion: both `std::shared_ptr` and `Lib::SmartPtr` have to allocate an additional control block somewhere, whereas you did this intrusively originally. If you can be bothered, `std::make_shared` ought to remove that overhead, but if not, the original was (is) fine by me, it just looked smartpointery.

Reply:
```
7344 ['problemsSTD_marton8472_lpo_dpc_i2000_2.pkl']
7352 ['problemsSTD_marton8472_lpo_dpc_i2000.pkl']
7354 ['problemsSTD_marton8473_lpo_dpc_i2000_2.pkl']
7355 ['problemsSTD_master8465_lpo_dpc_i2000.pkl']
7357 ['problemsSTD_marton8471_lpo_dpc_i2000_2.pkl']
7360 ['problemsSTD_marton8471_lpo_dpc_i2000.pkl']
7362 ['problemsSTD_marton8473_lpo_dpc_i2000.pkl']
```

intrude ~ 7358.5
Smart ~ 7358.0
shared ~ 7348.0

So I'd say let's keep Smart and that's it.

Reply:
@quickbeam123 Thanks for the comparison!

@MichaelRawson sorry for not taking the advice! I think we could have a `SmartPtr` removal PR at some point if this is really the case. Otherwise, I would vote for doing something "smart" as the original code looked terrible and it looks like it doesn't hurt much.

Reply:
I'm happy! Glad the nice thing worked out somehow. :smile: